src/services/refactors/extractType.ts

```diff
@@ -0,0 +1,141 @@
+        const selection = findAncestor(current, (node => node.parent && rangeContainsSkipTrivia(range, node, file) && !rangeContainsSkipTrivia(range, node.parent, file)));
+        if (!selection || !isTypeNode(selection)) return undefined;
+
+        if (isJS) {
+            // typeparam tag is not supported yet
+            return {
+                isJS,
+                selection,
+                firstStatement: Debug.assertDefined((findAncestor(selection, isStatementButNotBlockAndHasJSDoc))),
+            };
+        }
+        else {
+            const firstStatement = Debug.assertDefined((findAncestor(selection, isStatementButNotBlock)));
+            const typeParameters = checkAndCollectionTypeArguments(context.program.getTypeChecker(), selection, firstStatement, file);
+            if (!typeParameters) return undefined;
+
+            return { isJS, selection, firstStatement, typeParameters };
+        }
+    }
+
+    function isStatementButNotBlock(n: Node): n is Statement {
+        return n && isStatement(n) && !isBlock(n);
+    }
+
+    function isStatementButNotBlockAndHasJSDoc(n: Node): n is (Statement & HasJSDoc) {
+        return isStatementButNotBlock && hasJSDocNodes(n);
+    }
+
+    function rangeContainsSkipTrivia(r1: TextRange, node: Node, file: SourceFile): boolean {
+        return rangeContainsStartEnd(r1, skipTrivia(file.text, node.pos), node.end);
+    }
+
+    function checkAndCollectionTypeArguments(checker: TypeChecker, selection: TypeNode, statement: Statement, file: SourceFile): string[] | undefined {
+        let hasError = false;
+        const result: string[] = [];
+        visitor(selection);
+        return hasError ? undefined : result;
+
+        function visitor(node: Node) {
+            if (isTypeReferenceNode(node)) {
+                if (isIdentifier(node.typeName)) {
+                    const symbol = checker.resolveName(node.typeName.text, node.typeName, SymbolFlags.TypeParameter, /* excludeGlobals */ true);
+                    if (symbol) {
+                        const declaration = first(symbol.declarations);
+                        if (rangeContainsSkipTrivia(statement, declaration, file) && !rangeContainsSkipTrivia(selection, declaration, file)) {
+                            result.push(node.typeName.text);
+                        }
+                    }
+                }
+            }
+            else if (isInferTypeNode(node)) {
+                const conditionalTypeNode = findAncestor(node, isConditionalTypeNode);
+                if (!conditionalTypeNode || !rangeContainsSkipTrivia(selection, conditionalTypeNode, file)) {
+                    hasError = true;
+                    return;
+                }
+            }
+            else if (isTypePredicateNode(node) && !rangeContainsSkipTrivia(selection, node.parent, file)) {
+                hasError = true;
+                return;
+            }
+            else if (isTypeQueryNode(node) && isIdentifier(node.exprName)) {

```

Comment:
same as above, why is it not necessary to handle the case where `node.typeName` is not an identifier but is an entityName?

Reply:
usually, I think that `QualifiedName`  is not including in current statement

Reply:
And is not a TypeParameter and do not have TypeArguments

Reply:
like before, i think common cases are covered by the identifier check, but just came up with the following code:
```ts
interface I { a: string, f: (this: O, b: number) => typeof this.a };
```
something bad happens if we extract `typeof this.a`

Reply:
Banned `ThisTypeNode`, `QualityName` and most left is `ThisIdentifier`